doc/design/milestone-0.1.1/csv-generation.md

```diff
@@ -0,0 +1,162 @@
+
+1. Update an existing CSV at the currently pre-defined location, using available data in yaml manifests and source files.
+    
+    1. `renderCSV` will check for an existing CSV in `deploy`. Upon finding one, the CSV yaml file contents will be marshalled into a `ClusterServiceVersion` object.
+    1. Same as above.
+    1. Same as above.
+    
+### Extensible `CSVUpdater` CSV update mechanism
+
+The CSV spec will likely change over time as new Kubernetes and OLM features are implemented; we need the ability to easily extend the update system. The SDK will define the `CSVUpdater` interface as follows to encapsulate individual CSV field updates in methods:
+
+```Go
+import "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+
+type CSVUpdater interface {
+	Apply(*v1alpha1.ClusterServiceVersion) error
+}
+```
+
+`Apply` will use data from the `CSVUpdater` implementer to operate on `*v1alpha1.ClusterServiceVersion` object fields relevant to that updater. The OLM defines the entire CSV spec [in a Golang struct][olm-csv-spec-code] the SDK can alias to implement `CSVUpdater`s.
+
+Once sub-step two is reached when creating or updating a CSV, `renderCSV` will extract each yaml document discovered, and pass document data into a dispatcher function. The dispatcher selects the correct `CSVUpdater` to call based on the documents' `Kind` yaml object, a manifest type identifier used in all operator manifests.
+
+The following is an example implementation of an [install strategy][olm-csv-install-strat-doc] `CSVUpdater`:
+
+```Go
+import (
+  ...
+  appsv1 "k8s.io/api/apps/v1"
+  rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+  "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+  "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/install"
+)
+
+// CSVInstallStrategyUpdate embeds the OLM's install strategy spec.
+type CSVInstallStrategyUpdate struct {
+	*install.StrategyDetailsDeployment
+  
+  // Future utility fields go here.
+}
+
+// getLocalInstallStrategyCache retrieves the local cache singleton and returns the install strategy cache.
+func getLocalInstallStrategyCache() *CSVInstallStrategyUpdate {
+	factory := getLocalCacheFactory()
+	return factory.InstallStrategyCache
+}
+
+// AddDeploymentSpecToCSVInstallStrategyUpdate adds an RBAC Role to the local cache singletons' permissions.
+func AddRoleToCSVInstallStrategyUpdate(yamlDoc []byte) error {
+	localInstallStrategyUpdate := getLocalInstallStrategyCache()
+
+	newRBACRole := new(rbacv1beta1.Role)
+	_ = yaml.Unmarshal(yamlDoc, newRBACRole)
+
+	newPermissions := install.StrategyDeploymentPermissions{
+		ServiceAccountName: newRole.ObjectMeta.Name,
+		Rules:              newRole.Rules,
+	}
+	localInstallStrategyUpdate.Permissions = append(localInstallStrategyUpdate.Permissions, newPermissions)
```

Comment:
Same for the permissions section. In case `deploy/role.yaml` is updated we'll be overwriting the existing permissions based on that. Otherwise on each update you're appending the same permissions over and over.